src/deepsparse/v2/middleware/timer_middleware.py

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+import time
+from typing import Dict, Optional
+
+from deepsparse.v2.utils.state import InferenceState
+
+
+class Timer:
+    def __init__(self):
+        self._lock = threading.Lock()
+        self.measurements = {}
+        self.start_times = {}
+
+    def start(self, key: str):
+        """
+        Starts the timer
+        :param key: the key to track the timer for
+        """
+        with self._lock:
+            self.start_times[key] = time.time()
+
+    def end(self, key: str) -> bool:
+        """
+        Ends the timer and saves the runtime into measurements
+        :param key: the key to save the run time for
+        """
+        end_time = time.time()
+
+        with self._lock:
+            if key in self.start_times:
+                start_time = self.start_times[key]
+                del self.start_times[key]
+                self.measurements[key] = end_time - start_time
```

Comment:
let's make `self.measurements` a `Dict[str, List[float]]` so that we can keep multiple measurements for the same key (ie route splitting potentially)

Reply:
if we keep a list how do we know what recording is what